c/tskit/trees.c

```diff
@@ -2124,6 +2124,580 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+    tsk_safe_free(result_tmp);
+    tsk_bit_array_free(&ss_A_samples);
+    tsk_bit_array_free(&ss_B_samples);
+    tsk_bit_array_free(&ss_AB_samples);
+    tsk_bit_array_free(&AB_samples);
+    return ret;
+}
+
+static int
+get_all_mutation_samples(const tsk_treeseq_t *self, const tsk_size_t *site_offsets,
+    tsk_size_t *num_alleles, tsk_bit_array_t *allele_samples)
+{
+    int ret = 0;
+    tsk_tree_t tree;
+
+    ret = tsk_tree_init(&tree, self, TSK_TS_INIT_BUILD_INDEXES | TSK_NO_SAMPLE_COUNTS);
+    if (ret != 0) {
+        goto out;
+    }
+
+    for (ret = tsk_tree_first(&tree); ret == TSK_TREE_OK; ret = tsk_tree_next(&tree)) {
+        ret = get_mutation_samples(&tree, site_offsets, allele_samples, &num_alleles);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+    // if adding code below, check ret before continuing
+
+out:
+    tsk_tree_free(&tree);
+    return ret;
+}
+
+static int
+tsk_treeseq_two_site_count_stat(const tsk_treeseq_t *self, tsk_size_t state_dim,
+    const tsk_bit_array_t *sample_sets, tsk_size_t result_dim, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, norm_func_t *norm_f,
+    const double *TSK_UNUSED(left_window), const double *TSK_UNUSED(right_window),
+    tsk_flags_t options, tsk_size_t *result_size, double **result)
+{
+    int ret = 0;
+
+    tsk_bit_array_t all_samples_bits, allele_samples, allele_samples_row;
+    tsk_bit_array_t site_a_state, site_b_state;
+    tsk_size_t site_id, num_alleles_cumsum, inner, result_offset;
+    tsk_size_t site_a, site_b;
+    bool polarised = false;
+
+    const tsk_size_t num_sites = self->tables->sites.num_rows;
+    const tsk_size_t num_samples = self->num_samples;
+    const tsk_size_t max_alleles = self->tables->mutations.num_rows + num_sites;
+    tsk_size_t *restrict site_offsets = tsk_malloc(num_sites * sizeof(*site_offsets));
+    tsk_size_t *restrict num_alleles = tsk_malloc(num_sites * sizeof(*num_alleles));
+
+    tsk_memset(&all_samples_bits, 0, sizeof(all_samples_bits));
+    tsk_memset(&allele_samples, 0, sizeof(allele_samples));
+
+    if (site_offsets == NULL || num_alleles == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    ret = tsk_bit_array_init(&allele_samples, num_samples, max_alleles);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_bit_array_init(&all_samples_bits, num_samples, 1);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // A future improvement could get a union of all sample sets
+    // instead of all samples
+    get_all_samples_bits(&all_samples_bits, num_samples);
+
+    num_alleles_cumsum = 0;
+    for (site_id = 0; site_id < num_sites; site_id++) {
+        // Initialize the allele_samples with all samples in the ancestral allele
+        tsk_bit_array_get_row(&allele_samples, num_alleles_cumsum, &allele_samples_row);
+        tsk_bit_array_add(&allele_samples_row, &all_samples_bits);
+        // Store the allele offset for each site
```

Comment:
Do we need to do this stuff with ``site_offsets``, if we're iterating over the mutations using the inbuilt structs from the tree (as suggested above)?

Reply:
These offsets were put in place to reduce the amount of times I'd have to compute the offsets and to make sure all of the code is in sync. Now that things have been simplified and I'm not using offsets in 3 places, I think it's best to get rid of the malloc'd array of offsets. We still need offsets to index into the array of allele_samples, but now it's loop counters instead of a whole malloc'd array.